File: code/0110 Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        height = self.height(root)
        return height != -1
    
    def height(self, root):
        if not root:
            return 0
        left_depth = self.height(root.left)
        if left_depth == -1:
            return -1
        right_depth = self.height(root.right)
        if right_depth == -1:
            return -1
        if abs(left_depth - right_depth) > 1:
            return -1
        else:
            return max(left_depth, right_depth) + 1

File: code/0110 Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

#define MAX(a,b) ((a>b)?a:b)
int Balanced(struct TreeNode* root) {
    int left, right, diff;
    if (root==NULL) return 0;
    
    right = Balanced(root->right);
    left = Balanced(root->left);
    if (right == -1 || left == -1) return -1;
        
    diff = (right > left) ? right - left : left -right;
    if (diff > 1) return -1;
    else return 1 + MAX(left, right);
        
}

bool isBalanced(struct TreeNode* root) {
    return (Balanced(root) != -1);
}
